staging/src/k8s.io/kubectl/pkg/cmd/create/create_secret.go

```diff
@@ -91,233 +138,276 @@ func NewCmdCreateSecretGeneric(f cmdutil.Factory, ioStreams genericclioptions.IO
+		return fmt.Errorf("name must be specified")
+	}
+	if len(o.EnvFileSource) > 0 && (len(o.FileSources) > 0 || len(o.LiteralSources) > 0) {
+		return fmt.Errorf("from-env-file cannot be combined with from-file or from-literal")
+	}
+	return nil
 }
 
-// Complete completes all the required options
-func (o *SecretDockerRegistryOpts) Complete(f cmdutil.Factory, cmd *cobra.Command, args []string) error {
-	name, err := NameFromCommandArgs(cmd, args)
+// Run calls createSecret which will create secret based on CreateSecretOptions
+// and makes an API call to the server
+func (o *CreateSecretOptions) Run() error {
+	secret, err := o.createSecret()
 	if err != nil {
 		return err
 	}
-
-	fromFileFlag := cmdutil.GetFlagStringSlice(cmd, "from-file")
-	if len(fromFileFlag) == 0 {
-		requiredFlags := []string{"docker-username", "docker-password", "docker-server"}
-		for _, requiredFlag := range requiredFlags {
-			if value := cmdutil.GetFlagString(cmd, requiredFlag); len(value) == 0 {
-				return cmdutil.UsageErrorf(cmd, "flag %s is required", requiredFlag)
+	err = util.CreateOrUpdateAnnotation(o.CreateAnnotation, secret, scheme.DefaultJSONEncoder())
+	if err != nil {
+		return err
+	}
+	if o.DryRunStrategy != cmdutil.DryRunClient {
+		createOptions := metav1.CreateOptions{}
+		if o.FieldManager != "" {
+			createOptions.FieldManager = o.FieldManager
+		}
+		if o.DryRunStrategy == cmdutil.DryRunServer {
+			err := o.DryRunVerifier.HasSupport(secret.GroupVersionKind())
+			if err != nil {
+				return err
 			}
+			createOptions.DryRun = []string{metav1.DryRunAll}
+		}
+		secret, err = o.Client.Secrets(o.Namespace).Create(context.TODO(), secret, createOptions)
+		if err != nil {
+			return fmt.Errorf("failed to create secret %v", err)
 		}
 	}
 
-	var generator generate.StructuredGenerator
-	switch generatorName := cmdutil.GetFlagString(cmd, "generator"); generatorName {
-	case generateversioned.SecretForDockerRegistryV1GeneratorName:
-		generator = &generateversioned.SecretForDockerRegistryGeneratorV1{
-			Name:        name,
-			Username:    cmdutil.GetFlagString(cmd, "docker-username"),
-			Email:       cmdutil.GetFlagString(cmd, "docker-email"),
-			Password:    cmdutil.GetFlagString(cmd, "docker-password"),
-			Server:      cmdutil.GetFlagString(cmd, "docker-server"),
-			AppendHash:  cmdutil.GetFlagBool(cmd, "append-hash"),
-			FileSources: cmdutil.GetFlagStringSlice(cmd, "from-file"),
+	return o.PrintObj(secret)
+}
+
+// createSecret fills in key value pair from the information given in
+// CreateSecretOptions into *corev1.Secret
+func (o *CreateSecretOptions) createSecret() (*corev1.Secret, error) {
+	namespace := ""
+	if o.EnforceNamespace {
+		namespace = o.Namespace
+	}
+	secret := &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: corev1.SchemeGroupVersion.String(),
+			Kind:       "Secret",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      o.Name,
+			Namespace: namespace,
+		},
+	}
+	secret.Data = map[string][]byte{}
```

Comment:
Here the code can also make some use of the helper function stated below (in the create_secret_docker comments).

The function here can be called with newSecretObj(name, namespace, corev1.SecretTypeOpaque) and then, keep the code below to change only if type is defined